libvast/src/system/disk_monitor.cpp

```diff
@@ -78,11 +80,12 @@ disk_monitor(disk_monitor_actor::stateful_pointer<disk_monitor_state> self,
       // have finished or we encountered an error.
       auto shared_guard
         = make_shared_guard([=] { self->state.purging = false; });
-      directory index_dir = self->state.dbdir / "index";
+      auto index_dir
+        = std::filesystem::directory_iterator(self->state.dbdir / "index");
```

Comment:
Please switch to the error code version here, change the return type of the lambda to `caf::result<void>`, and return an error created using `caf::make_error(ec::filesystem_error, ...)` when the the call failed. All other places where this lambda returned normally can be replaced with `return {};`.

Reply:
The error can then be received on the callsite like this:

Replace:

```
self->send(disk_monitor, atom::erase_v)
``` 

with

```
self->request(disk_monitor, caf::infinite, atom::erase_v).then(
  []() {
    // nop, all good
  },
  [=](const caf::error& err) {
    VAST_ERROR("{} failed to ...: {}`, self, err);
  });
```

Reply:
I switched to the `error_code` overload, but I had some issues making the call sites within `disk_monitor.cpp` work with your suggestion. I tried replacing `disk_monitor` with `self->state.db_dir` in the two call sites, but that didn't quite work as the handle. Can you elaborate a bit more on changing the `send` calls to `request` with appropriate error handling?

By the way, to contribute to my branch, it should suffice to do something like:

```
git remote add joe git@github.com:JoeLoser/vast.git
git fetch joe
git checkout -b recursive-size joe/recursive-size
// make changes and commit
git push origin joe/recursive-size
```

I don't mind if you push onto my branch when it makes sense as it may be easier than explaining the code over PR.

Reply:
I pushed the necessary changes. `self` is a stateful pointer to the disk monitor actor, and due to a deficiency in the CAF version we're using (fixed upstream already), the pointer must be dereferenced manually to specify the receiving actor when using `request` over `send`. That you really could not know. 

Fwiw I had to reset my local VAST repository to add the remote successfully, which is probably why I thought I couldn't push last time around.

Reply:
Also, if you ever want a quick intro on working with CAF, feel free to just ask.

Reply:
Thanks for pushing the commit -- I left two small comments regarding the changes. Glad you got the remote set up -- I suspect it'll be useful in the future 👍 

As for the CAF intro, I'll probably take you up on that in the future once I get a bit more understanding of the codebase and how Vast uses it. So far, it seems to rely on the actors a fair bit.

Reply:
Turns out I actually had to static_cast to the typed actor handle type. The switch to CAF 0.18 cannot come soon enough... If only that refactoring wasn't as big of a change and not a breaking change. 🙄

Reply:
Glad you spotted the error quickly -- I just noticed that when building locally and the CI jobs failing as well.